src/libopensc/pkcs15-piv.c

```diff
@@ -1200,6 +1220,61 @@ sc_log(card->ctx,  "DEE Adding pin %d label=%s",i, label);
 			LOG_FUNC_RETURN(card->ctx, r);
 	}
 
+	/* set pins */
+	sc_log(card->ctx,  "PIV-II adding pins...");
+	for (i = 0; pins[i].label; i++) {
+		struct sc_pkcs15_auth_info pin_info;
+		struct sc_pkcs15_object   pin_obj;
+		char label[SC_PKCS15_MAX_LABEL_SIZE -1];
+		int pin_ref;
+
+		memset(&pin_info, 0, sizeof(pin_info));
+		memset(&pin_obj,  0, sizeof(pin_obj));
+
+		pin_info.auth_type = SC_PKCS15_PIN_AUTH_TYPE_PIN;
+		sc_pkcs15_format_id(pins[i].id, &pin_info.auth_id);
+		pin_info.attrs.pin.reference     = pins[i].ref;
+		pin_info.attrs.pin.flags         = pins[i].flags;
+		pin_info.attrs.pin.type          = pins[i].type;
+		pin_info.attrs.pin.min_length    = pins[i].minlen;
+		pin_info.attrs.pin.stored_length = pins[i].storedlen;
+		pin_info.attrs.pin.max_length    = pins[i].maxlen;
+		pin_info.attrs.pin.pad_char      = pins[i].pad_char;
+		sc_format_path(pins[i].path, &pin_info.path);
+		pin_info.tries_left    = -1;
+
+		strncpy(label, pins[i].label, SC_PKCS15_MAX_LABEL_SIZE -1);
+		if (i == 0 &&
+			sc_card_ctl(card, SC_CARDCTL_PIV_PIN_PREFERENCE,
+					&pin_ref) == 0 &&
+				pin_ref == 0x00) { /* must be 80 for PIV pin, or 00 for Global PIN */
+			pin_info.attrs.pin.reference = pin_ref;
+			pin_info.attrs.pin.flags &= ~SC_PKCS15_PIN_FLAG_LOCAL;
+			strncpy(label, "Global PIN", SC_PKCS15_MAX_LABEL_SIZE -1);
+		}
+		if (has_yubikey_touch_policy) {
+			int offset = strlen(label);
+			strncpy(label + offset, ",Yubikey has touch policy", SC_PKCS15_MAX_LABEL_SIZE - offset - 1);
+		}
```

Comment:
I think there are some corner cases when copying the label around.

since `label` is uninitialized it may not contain a termination. When you're copying the label (line 1246), this is still the case if the label is completely filled, because you're omitting the label's termination. This will then lead to a bad calculation of the length of `label`. Also, the calculation of index and lenght looks fishy.

Why don't you use strlcat?